### PR TITLE
Minimize the use of shared_ptr

### PR DESCRIPTION
**Pull request by bowu**
This commit minimizes the use of shared_ptr. In addition, it introduces three things.

1. ExecutionContext class (to be extended), which contains all needed runtime information to run an ExecutionPlan. 
2. OpGraph prepares all its ops by passing to them a data graph. 
3. A simple runner for ExecutionPlan.

### CODE REVIEW

File: libquery/include/querying/OpGraph.h [hunk 1]
```diff
@@ -8,31 +8,35 @@
...
 
 using AttributedGraph = graphs::AttributedGraph;
+class ExecutionPlan;
 
 class OpGraph {
 public:
   using OpPtr = std::shared_ptr<Operator>;
 
-  OpGraph() = default;
+  void SetOwner(std::shared_ptr<ExecutionPlan> plan) { owner_ = plan; }
```
Comment (roshandathathri) [inline libquery/include/querying/OpGraph.h:17]: What's the purpose for this?

File: libquery/include/querying/ops/MatchOp.h [hunk 2]
```diff
@@ -28,6 +28,9 @@ class MatchOp : public Operator {
   inline bool SanityCheck() final { return true; }
 
 private:
+  //! Data graph representing the input data graph to the query
+  AttributedGraph data_graph_;
```
Comment (roshandathathri) [inline libquery/include/querying/ops/MatchOp.h:32]: Why is this needed?

File: libquery/include/querying/ops/Operator.h [hunk 3]
```diff
@@ -94,6 +101,8 @@ class Operator : public std::enable_shared_from_this<Operator> {
     children_.push_back(child);
   }
 
+  std::shared_ptr<AttributedGraph> data_graph_;
```
Comment (roshandathathri) [inline libquery/include/querying/ops/Operator.h:104]: Why is this needed?

File: libquery/include/querying/OpGraph.h (hunk 1 above)
Comment (bowu) [reply libquery/include/querying/OpGraph.h:17]: The OpGraph can access the execution context through its owner, an execution plan.

File: libquery/include/querying/ops/Operator.h (hunk 3 above)
Comment (bowu) [reply libquery/include/querying/ops/Operator.h:104]: If we don't have this, every time an op would like to access the data graph it has to obtain it from the execution context.

File: libquery/include/querying/ops/MatchOp.h (hunk 2 above)
Comment (bowu) [reply libquery/include/querying/ops/MatchOp.h:32]: Good catch. This is not needed, as MatchOp inherits a data graph from Operator. I'll remove it.

File: libquery/include/querying/OpGraph.h (hunk 1 above)
Comment (roshandathathri) [reply libquery/include/querying/OpGraph.h:17]: Why does OpGraph need to access the plan? It is sufficient for the plan to access the OpGraph right?

File: libquery/include/querying/ops/Operator.h (hunk 3 above)
Comment (roshandathathri) [reply libquery/include/querying/ops/Operator.h:104]: Data graph will be accessed during execution right?

File: libquery/include/querying/OpGraph.h (hunk 1 above)
Comment (bowu) [reply libquery/include/querying/OpGraph.h:17]: This is just in case we need more than the data graph, but I can change the code to pass a pointer to OpGraph to access the execution context.

File: libquery/include/querying/ops/Operator.h (hunk 3 above)
Comment (bowu) [reply libquery/include/querying/ops/Operator.h:104]: That's correct.

**Comment by bowu** [comment]
Does anyone know why the Shared check fails but the Release check succeeds?

File: libquery/include/querying/OpGraph.h [hunk 4]
```diff
@@ -8,31 +8,35 @@
...
 
-  //! reset the graph for a new execution
+  //! reset the graph for a new round of execution
   void Reset() {}
 
   bool IsReady() { return true; }
 
 private:
   std::vector<OpPtr> ops_;
+  std::shared_ptr<AttributedGraph> data_graph_{nullptr};
```
Comment (ddn0) [inline libquery/include/querying/OpGraph.h:38]: FWIW, I think we should be wary of `shared_ptr` everywhere. It obscures when objects actually stopped being referenced.

In general it would be better to just have a clear container object like the plan, and then objects can be tied to the lifetime of the container object and they can simply borrow references (T*) to parts of the plan.

File: libquery/include/querying/QueryServer.h [hunk 5]
```diff
@@ -18,17 +18,18 @@ class QueryServer {
...
   ~QueryServer() = default;
 
-  Table RunCypherQuery(std::string graph_file_name, std::string query_str) {
+  std::unique_ptr<Table> RunCypherQuery(std::string graph_file_name,
+                                        std::string query_str) {
     std::shared_ptr<AttributedGraph> graph = LoadPropertyGraph(graph_file_name);
-    // TODO(bowu) plan/query graph is tied to some schema; query_str is not
-    // enough, need graph schema to create it; tie graph schema to cache as well
-    std::shared_ptr<ExecutionPlan> plan = cache_.GetPlan(query_str);
+    // TODO: this is a temporary fix; graph may change to make the plan invalid
```
Comment (ddn0) [inline libquery/include/querying/QueryServer.h:24]: probably keep your name associated with TODOs.

File: libquery/include/querying/ops/Operator.h [hunk 6]
```diff
@@ -7,9 +7,12 @@
 
 #include "querying/InterData.h"
 #include "galois/Logging.h"
+#include "galois/graphs/AttributedGraph.h"
 
 namespace galois {
 
+using AttributedGraph = graphs::AttributedGraph;
```
Comment (ddn0) [inline libquery/include/querying/ops/Operator.h:14]: ditto

File: libquery/include/querying/ops/Operator.h [hunk 7]
```diff
@@ -7,9 +7,12 @@
...
+
 class Operator : public std::enable_shared_from_this<Operator> {
```
Comment (ddn0) [inline libquery/include/querying/ops/Operator.h:16]: For graph-like data structures, I wouldn't expect reference counting to be a great fit (even if you can assert that it will be a DAG). I'd opt for a container that owns all the Ops in a graph.

File: libquery/CMakeLists.txt [hunk 8]
```diff
@@ -16,6 +16,7 @@ target_sources(galois_query PRIVATE
   src/Matching.cpp
   src/RegularPathQuery.cpp
   src/SubgraphQuery.cpp
+  src/OpGraph.cpp
```
Comment (ddn0) [inline libquery/CMakeLists.txt:19]: small nit: just sort this list and keep insertions sorted.

File: libquery/include/querying/OpGraph.h (hunk 4 above)
Comment (l-hoang) [reply libquery/include/querying/OpGraph.h:38]: In the case of the data_graph_, having a unique pointer for the Context (the owner) and having the children of Context keep a pointer to it is what I would suggest.

File: libquery/src/OpGraph.cpp [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace galois {
+
+// TODO: Do we need a Prepare function for each Op?
+void OpGraph::Prepare() {
+  std::shared_ptr<ExecutionPlan> plan = owner_.lock();
+  if (!plan) {
+    GALOIS_LOG_FATAL("OpGraph does not have an owner");
+  }
+  auto dataGraph = plan->GetExecutionContext().GetDataGraph();
```
Comment (ddn0) [inline libquery/src/OpGraph.cpp:12]: `data_graph`

File: libquery/include/querying/ExecutionPlan.h [hunk 10]
```diff
@@ -1,37 +1,36 @@
...
-using AttributedGraph = graphs::AttributedGraph;
-
-class ExecutionPlan {
+class ExecutionPlan : public std::enable_shared_from_this<ExecutionPlan> {
 
 public:
-  ExecutionPlan(std::string query_str,
-                const galois::BitLabelSchema& bit_label_schema)
-      : query_str_(query_str) {
+  ExecutionPlan(ExecutionContext ctx) : ctx_(ctx) {
```
Comment (l-hoang) [inline libquery/include/querying/ExecutionPlan.h:18]: Pass this in via reference to avoid a copy. `ExecutionContext&`

File: libquery/include/querying/ExecutionPlan.h [hunk 11]
```diff
@@ -59,11 +58,9 @@ class ExecutionPlan {
     }
   }
 
-  std::string query_str_;
+  ExecutionContext ctx_;
```
Comment (l-hoang) [inline libquery/include/querying/ExecutionPlan.h:61]: Make this a reference, though my assumption here is that the context object is being kept live somewhere.

My guess is that you will reuse the context in many plans though.

File: libquery/include/querying/ExecutionContext.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      : query_str_(query_str), graph_(graph), opt_level_(opt_level) {}
+
+  std::string GetQueryStr() { return query_str_; }
+
+  std::shared_ptr<AttributedGraph> GetDataGraph() { return graph_; }
+
+  unsigned int GetOptLevel() { return opt_level_; }
+
+private:
+  std::string query_str_;
```
Comment (l-hoang) [inline libquery/include/querying/ExecutionContext.h:28]: Each context has its own query?

File: libquery/CMakeLists.txt (hunk 8 above)
Comment (l-hoang) [reply libquery/CMakeLists.txt:19]: Is there an automated way to do this sorting or will it have to be good ol' eyeballing and line moving?

File: libquery/include/querying/ExecutionContext.h [hunk 13]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <string>
+
+#include "galois/graphs/AttributedGraph.h"
+
+namespace galois {
+
+using AttributedGraph = graphs::AttributedGraph;
```
[the remark this answers is not in the document]
Comment (bowu) [reply libquery/include/querying/ExecutionContext.h:9]: Will do.

File: libquery/include/querying/OpGraph.h (hunk 4 above)
Comment (bowu) [reply libquery/include/querying/OpGraph.h:38]: Make sense.

File: libquery/CMakeLists.txt (hunk 8 above)
Comment (ddn0) [reply libquery/CMakeLists.txt:19]: in vim, I do a visual select followed by `:!sort`

File: libquery/include/querying/OpGraph.h (hunk 4 above)
Comment (bowu) [reply libquery/include/querying/OpGraph.h:38]: The reason that I used shared_ptr is because I wanted to avoid two indirections to get it from an ExecutionContext. We can revisit this later.

Comment (l-hoang) [reply libquery/include/querying/OpGraph.h:38]: I think what we're saying is that you don't need to keep it a shared_ptr but rather a regular pointer AttributedGraph; it  avoids the additional reference increment to the shared pointer.

Comment (bowu) [reply libquery/include/querying/OpGraph.h:38]: Got it.

**Comment by bowu** [comment]
@ddn0 Can you take a look at the status checks? It seems a mac os check failure stopped the world.

File: libquery/include/querying/OpGraph.h [hunk 14]
```diff
@@ -4,35 +4,37 @@
 
 #include "querying/ops/Operators.h"
 #include "galois/graphs/AttributedGraph.h"
+#include "querying/Table.h"
 
 namespace galois {
 
-using AttributedGraph = graphs::AttributedGraph;
+class ExecutionContext;
```
Comment (l-hoang) [inline libquery/include/querying/OpGraph.h:11]: Prefer to include the header to the Context class rather than declare a forward reference.

File: libquery/include/querying/ops/FilterOp.h [hunk 15]
```diff
@@ -6,7 +6,7 @@ namespace galois {
 
 class FilterOp : public Operator {
 public:
-  virtual void Compute() final {}
+  std::vector<Operator*> Compute() final { return PropagateOutput(); }
```
Comment (l-hoang) [inline libquery/include/querying/ops/FilterOp.h:9]: I'm assuming this is dummy work for now (as in this isn't the final Compute function).

File: libquery/include/querying/ops/Operator.h [hunk 16]
```diff
@@ -73,39 +75,56 @@ class Operator : public std::enable_shared_from_this<Operator> {
...
-        std::cerr << "Child has been deallocated\n";
-        abort();
+      child->TakeInput(output_);
+      if (child->IsReady()) {
+        readyOps.emplace_back(child);
       }
     }
+    // clear the output if it's not a terminal op
+    if (NumChildren() != 0) {
+      output_ = nullptr;
```
Comment (l-hoang) [inline libquery/include/querying/ops/Operator.h:122]: output_.reset() is an explicit release of the pointer and makes it clear that this is a shared pointer.
